# Suma de los primeros 1 000 000 000 números naturales

Para calcular la suma de los primeros $n$ números naturales se utiliza la formula $\frac{n(n+1)}{2}$.

Con fines didacticos para el curso de Algoritmos Paralelos se diseño una práctica donde se implementa la suma de los primeros 1 000 000 000 haciendo uso de hilos. En la siguiente sección se presenta la forma secuencial de hacer la suma, la optimización secuencial usando el decorador *@jit* de la libreria *Numba* y calculando con la formula la suma.

In [1]:
int((100*(100+1))/2)

5050

Al ser una operación tiene una complejidad $O(1)$ respecto a tiempo y espacio en memoria.

In [2]:
def suma100numeros(x):
    suma = 0
    for i in range(0,101):
        suma+=i
    if x == 1:
        print("La suma de los primeros 100 es: ",suma)

La complejidad en tiempo de la función *suma100numeros* tiene complejidad $O(n)$ donde *n* es el número limite de la suma. La complejidad en espacio es constante $O(1)$ pues la función *suma100numeros* hace uso exclusivo de una variable donde se guarda la suma en cada iteración.

In [3]:
suma100numeros(1)

La suma de los primeros 100 es:  5050


In [4]:
%timeit suma100numeros(0)

4.27 µs ± 130 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


Usando la fórmula para calcular la suma de los primeros 1 000 000 000 números naturales

In [5]:
((1000000000*(1000000000+1))/2)

5.000000005e+17

//////Tipo de entero que permita visualizar? 

In [6]:
def sumaNumeros():
    suma = 0
    for i in range(0,100000001):
        suma+=i
    #print("La suma de los primeros 1 000 000 000 es: ",suma)

In [14]:
%timeit sumaNumeros()

KeyboardInterrupt: 

Observemos que calcular la suma de los primeros $100$ números tiene un costo de $4.35 \mu s$ mientras que calcular la suma de los primeros $1 \: 000 \: 000 \: 000$ tiene un costo de $5.76$ segundos 
>¿Sería bueno mencionar cuanto aumenta proporcionalmente?

Ahora haremos las mismas pruebas usando la optimización que nos brinda *numba*

In [21]:
from numba import jit, prange

In [10]:
@jit
def suma100numerosJIT(x):
    suma = 0
    for i in range(0,101):
        suma+=i
    if x == 1:
        print("La suma de los primeros 100 es: ",suma)

In [11]:
%timeit suma100numerosJIT(0)

The slowest run took 13.24 times longer than the fastest. This could mean that an intermediate result is being cached.
897 ns ± 1.3 µs per loop (mean ± std. dev. of 7 runs, 1 loop each)


In [12]:
@jit
def sumaNumerosJIT():
    suma = 0
    for i in range(0,100000001):
        suma+=i
    #print("La suma de los primeros 1 000 000 000 es: ",suma)

In [13]:
%timeit sumaNumerosJIT()

67.6 ns ± 0.258 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [ ]:
## Para obtener información desde Jupyter sobre el SO
##!numba -s


Con las pruebas en tiempo de los mismos métodos pero optimizados con el decorador *@jit* obtenemos que sumar los primeros cien números tiene un costo de $161ns$ (nanosegundos) y la suma de los primeros mil millones de números naturales  tiene un costo de $65.4 ns$ (nanosegundos), simplemente a diferencia de los metodos secuenciales sin optimizar coloca en el mismo renglon el costo de ambos procesos. 
> Sin embargo, la optimización del decorador @jit y el uso de Numba puede no quedar pertinente en esta sección más que para hacer una comparación con el uso de hilos en paralelos ya que parece ser que la optimización de numba usa el tipo de variables para eficientar el codigo y no el paralelismo. Hay que investigar más. 

Valor | Símbolo | Nombre
:----: | ------ | ------
    10⁻¹s | ds | decisegundos
    10⁻²s | cs  | centisegundos
    10⁻³s | ms | milisegundos
    10⁻⁶s  | $\mu s$ |microsegundos
    10⁻⁹ | ns | nanosegundos 
    10⁻¹² | ps | spicosegundos

Ahora paralelizare la función además de optimizarla con el decorador *@jit*

In [32]:
@jit(parallel=True)
def sumaNumerosJITP():
    suma = 0
    for i in prange(0,100000001):
        suma+=i
    #print("La suma de los primeros 1 000 000 000 es: ",suma)

In [33]:
%timeit sumaNumerosJITP

22.7 ns ± 2.42 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)


La prueba realizada nos da como resultado un tiempo de $18.8 ns$, sin paralelizar el proceso tomaba entre $67.2$ nanosegundos con un eror de mas menos $0.258$ nanosegundos. La indemiata desventaja que encuentro en esta forma de paralelizar es que el alumno no decide como subdividir las tareas del problema, sino que numba hace el trabajo al indicarle el parametro *parallel* con *True*. Tenemos una mejora de hasta $49$ nanosegundos.

In [34]:
sumaNumerosJITP()
sumaNumerosJITP.parallel_diagnostics(level=4)

 
 Parallel Accelerator Optimizing:  Function sumaNumerosJITP, <ipython-
input-32-1fe85ec8da77> (1)  


Parallel loop listing for  Function sumaNumerosJITP, <ipython-input-32-1fe85ec8da77> (1) 
-------------------------------------|loop #ID
@jit(parallel=True)                  | 
def sumaNumerosJITP():               | 
    suma = 0                         | 
    for i in prange(0,100000001):----| #0
        suma+=i                      | 
--------------------------------- Fusing loops ---------------------------------
Attempting fusion of parallel loops (combines loops with similar properties)...
----------------------------- Before Optimisation ------------------------------
--------------------------------------------------------------------------------
------------------------------ After Optimisation ------------------------------
Parallel structure is already optimal.
--------------------------------------------------------------------------------
---------------------------------